### PR TITLE
Test:Refactoring lib/plugins/aws/package/lib/mergeIamTemplates.test.js 

### PR DESCRIPTION
**Pull request by caiobessa**
Refactoring `lib/plugins/aws/package/lib/mergeIamTemplates.test.js` to use runServerless based variant.

Closes: #8524

### CODE REVIEW

**Review by medikoo — COMMENTED** [review]
@calobessa Thank you! Do you know why CI operations are shown as _cancelled_? We need a full CI run to confirm it doesn't break

**Comment by caiobessa** [comment]
> @caiobessa Thank you! Do you know why CI operations are shown as _cancelled_? We need a full CI run to confirm it doesn't break

@medikoo there was a minor error in the code. I fixed it in the [683f023](https://github.com/serverless/serverless/pull/8575/commits/1969ccdb03a372971ac24c022438803f0683f023). Thanks!

File: test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js [hunk 1]
```diff
@@ -3,865 +3,267 @@
...
+describe('lib/plugins/aws/package/lib/mergeIamTemplates.test.js', () => {
+  describe('No default role', () => {
+    it('should not create role resource if there are no functions', async () => {
+      const { cfTemplate, awsNaming } = await runServerless({
+        fixture: 'aws',
+        cliArgs: ['package'],
       });
+      const iamRoleLambdaExecution = awsNaming.getRoleLogicalId();
+      const resourceIam = cfTemplate.Resources[iamRoleLambdaExecution];
+      expect(resourceIam).to.be.undefined;
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js:15]: Let's reconfigure to `expect(cfTemplate.Resources).to.not.have.property(iamRoleLambdaExecution)`, it's more up to the point

File: test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js [hunk 2]
```diff
@@ -3,865 +3,267 @@
...
-        ]
-      ).to.deep.equal({
-        Type: 'AWS::Logs::LogGroup',
-        Properties: {
-          LogGroupName: awsPackage.provider.naming.getLogGroupName(f.func1.name),
-        },
-      });
+      const IamRoleLambdaExecution = awsNaming.getRoleLogicalId();
+      const resourceIam = cfTemplate.Resources[IamRoleLambdaExecution];
+      expect(resourceIam).to.be.undefined;
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js:32]: Same here (and all other places in the file)

File: test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js [hunk 3]
```diff
@@ -3,865 +3,267 @@
...
+                handler: 'index.handler',
+              },
+              myFunctionWithRole: {
+                name: 'myCustomName',
+                handler: 'index.handler',
+                role: 'myCustRole0',
+              },
+            },
           },
+        }).then(({ cfTemplate, awsNaming }) => {
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js:82]: Let's stick to just async/await syntax (no `.then` calls), here and in all other places

File: test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js [hunk 4]
```diff
@@ -3,865 +3,267 @@
...
       it('should not configure ManagedPolicyArns by default', () => {
         // Replaces
-        // https://github.com/serverless/serverless/blob/d8527d8b57e7e5f0b94ba704d9f53adb34298d99/lib/plugins/aws/package/lib/mergeIamTemplates.test.js#L665-L683
+        const IamRoleLambdaExecution = naming.getRoleLogicalId();
+        const { Properties } = cfResources[IamRoleLambdaExecution];
+        expect(Properties.ManagedPolicyArns).to.be.undefined;
       });
 
       it('should add logGroup access policies if there are functions', () => {
         // Replaces
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js:96]: Let's remove that comment

File: test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js [hunk 5]
```diff
@@ -3,865 +3,267 @@
...
-        // Replaces
-        // https://github.com/serverless/serverless/blob/d8527d8b57e7e5f0b94ba704d9f53adb34298d99/lib/plugins/aws/package/lib/mergeIamTemplates.test.js#L412-L428
-        // https://github.com/serverless/serverless/blob/d8527d8b57e7e5f0b94ba704d9f53adb34298d99/lib/plugins/aws/package/lib/mergeIamTemplates.test.js#L492-L508
-        // https://github.com/serverless/serverless/blob/d8527d8b57e7e5f0b94ba704d9f53adb34298d99/lib/plugins/aws/package/lib/mergeIamTemplates.test.js#L510-L526
-      });
+      it('should support `provider.iamRoleStatements`', async () => {
+        const {
+          cfTemplate: { Resources },
+          awsNaming,
+        } = await runServerless({
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js:195]: There should be no `runServerless` run here. It should be tested through service setup in `before` clause

File: test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js [hunk 6]
```diff
@@ -880,6 +282,14 @@ describe('lib/plugins/aws/package/lib/mergeIamTemplates.test.js', () => {
                 handler: 'index.handler',
                 disableLogs: true,
               },
+              functionWithVpc: {
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js:285]: Let's follow same naming convention (name functions with `fn` prefix not `function` prefix)

**Review by medikoo — CHANGES_REQUESTED** [review]
@caiobessa that looks great! Really outstanding work. I've proposed just few improvements (mostly about conventions we're intended to follow), and it should be good to merge

**Comment by caiobessa** [comment]
> @caiobessa that looks great! Really outstanding work. I've proposed just few improvements (mostly about conventions we're intended to follow), and it should be good to merge

@medikoo Thanks a lot! I did request changes.

File: test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js [hunk 7]
```diff
@@ -3,865 +3,248 @@
...
-                    ],
-                  },
-                ],
-              },
+    it('should not create role resource with all functions having `functions[].role`', async () => {
+      const { cfTemplate, awsNaming } = await runServerless({
+        fixture: 'function',
+        cliArgs: ['package'],
+        configExt: {
+          service: 'another-service',
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js:38]: Let's not override this property (here in all other places)

File: test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js [hunk 8]
```diff
@@ -3,865 +3,248 @@
...
+          functions: {
+            foo: {
+              handler: 'index.handler',
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js:41]: We don't need to redefine handler (here and in all other places)

File: test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js [hunk 9]
```diff
@@ -880,6 +263,14 @@ describe('lib/plugins/aws/package/lib/mergeIamTemplates.test.js', () => {
                 handler: 'index.handler',
                 disableLogs: true,
               },
+              fnWithVpc: {
+                handler: 'func.function.handler',
+                name: 'new-service-dev-func1',
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js:268]: Do we need to configure this `name` property?

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @caiobessa ! It looks very good. I have just few final cleanup suggestions

File: test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js (hunk 9 above)
Comment (caiobessa) [reply test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js]: We don't I removed it

**Comment by caiobessa** [comment]
Hi @medikoo thanks for the suggestions.

File: test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js [hunk 10]
```diff
@@ -3,865 +3,244 @@
...
-    describe.skip('TODO: Defaults', () => {
       before(async () => {
-        await runServerless({
+        const { cfTemplate, awsNaming } = await runServerless({
           fixture: 'function',
+          cliArgs: ['package'],
           configExt: {
-            // Configure one of the functions with custom "role", that will replace:
-            // https://github.com/serverless/serverless/blob/d8527d8b57e7e5f0b94ba704d9f53adb34298d99/lib/plugins/aws/package/lib/mergeIamTemplates.test.js#L587-L611
+            service: 'my-service',
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js:160]: Let's remove this property

File: test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js [hunk 11]
```diff
@@ -880,6 +259,13 @@ describe('lib/plugins/aws/package/lib/mergeIamTemplates.test.js', () => {
                 handler: 'index.handler',
                 disableLogs: true,
               },
+              fnWithVpc: {
+                handler: 'func.function.handler',
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js:263]: Doesn't that break? I think this handler is pointing not existing file. Why it's not just `index.handler` (in context of this test fact that all lambdas point same files is not that relevant)

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @caiobessa ! We're near there, I have just final comments and we should be good to go

**Comment by caiobessa** [comment]
@medikoo Thanks for the review, if have anythings else feel free to let me know.

**Review by medikoo — APPROVED** [review]
Thank you @caiobessa !

**Pull request merged**